### PR TITLE
7033 signup message

### PR DESCRIPTION
**Pull request by TeroOtavia**
Changes:
- Workspace can now have custom signup message
- Workspace signup groups can have now custom signup message that exceeds workspace signup message if it was set.
- Workspace management refactored to be function component with multiple sub components to decrease re-rendering when state updates.
- Workspace permissions aka "signup groups" are now updated with one api call. This also prevents permission list to suffle items between updates.

Resolves: #7033

### CODE REVIEW

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 1]
```diff
@@ -1150,6 +1171,253 @@ public Response updateWorkspace(@PathParam ("WORKSPACEENTITYID") Long workspaceE
...
+    }
+
+    if (!workspaceController.canIManageWorkspaceSettings(workspaceEntity)) {
+      return Response.status(Status.FORBIDDEN).build();
+    }
+
+    WorkspaceSignupMessage defaultSignupMessage = workspaceSignupMessageController.findDefaultSignupMessage(workspaceEntity);
+    WorkspaceSignupMessageRestModel restModel = defaultSignupMessage != null
+        ? new WorkspaceSignupMessageRestModel(defaultSignupMessage.isEnabled(), defaultSignupMessage.getCaption(), defaultSignupMessage.getContent()) 
+        : new WorkspaceSignupMessageRestModel(false, "", "");
```
[the remark this answers is not in the document]
Comment (aviljakaine) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java:1190]: There were some front end issues with this and it was requested as empty strings for front end purposes. We did discuss if some other response code would've been more correct when the message doesn't yet exist. Ended up with this though.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 2]
```diff
@@ -1150,6 +1171,253 @@ public Response updateWorkspace(@PathParam ("WORKSPACEENTITYID") Long workspaceE
...
+      return Response.status(Status.NOT_FOUND).entity(String.format("WorkspaceEntity #%d not found", workspaceEntityId)).build();
+    }
+
+    if (!workspaceController.canIManageWorkspaceSettings(workspaceEntity)) {
+      return Response.status(Status.FORBIDDEN).build();
+    }
+
+    // Make sure the basic properties are defined
+    
+    if (payload == null || StringUtils.isAnyBlank(payload.getCaption(), payload.getContent())) {
```
[the remark this answers is not in the document]
Comment (aviljakaine) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java:1210]: No way to remove one, they can always be disabled though.

**Comment by aviljakaine** [comment]
> Couple of generic comments, and the course signup test might need adjusting?

Test fixed, the sent message from course picker signup was also changed a bit as it had a weird old part to it.

**Review by avukkonen — CHANGES_REQUESTED** [review]

**Review by avukkonen — APPROVED** [review]

**Review by pkukkon — APPROVED** [review]
Backend OK.

**Pull request merged**